### PR TITLE
refactor: simplify task state storing and restoring

### PR DESCRIPTION
**Pull request by montekki**

### CODE REVIEW

File: insonmnia/worker/server.go [hunk 1]
```diff
@@ -371,18 +372,38 @@ func (m *Worker) cancelDealTasks(deal *pb.Deal) error {
...
-	Spec        pb.TaskSpec   `json:"spec,omitempty"`
+	Description descriptionMarshallable   `json:"description,omitempty"`
+	Cinfo       containerInfoMarshallable `json:"cinfo,omitempty"`
+	Spec        pb.TaskSpec               `json:"spec,omitempty"`
+}
+
+type descriptionMarshallable struct {
+	Description
+	Reference reference.Field `json:"Reference"`
+	Networks  []*structs.NetworkSpec
```
Comment (antmat) [inline insonmnia/worker/server.go]: why can't we just make networks in description public?

File: insonmnia/worker/server.go [hunk 2]
```diff
@@ -371,18 +372,38 @@ func (m *Worker) cancelDealTasks(deal *pb.Deal) error {
...
-	Cinfo       ContainerInfo `json:"cinfo,omitempty"`
-	Spec        pb.TaskSpec   `json:"spec,omitempty"`
+	Description descriptionMarshallable   `json:"description,omitempty"`
+	Cinfo       containerInfoMarshallable `json:"cinfo,omitempty"`
+	Spec        pb.TaskSpec               `json:"spec,omitempty"`
+}
+
+type descriptionMarshallable struct {
+	Description
+	Reference reference.Field `json:"Reference"`
```
Comment (antmat) [inline insonmnia/worker/server.go]: maybe just store reference.Field in description for simplicity?

File: insonmnia/worker/server.go [hunk 3]
```diff
@@ -371,18 +372,38 @@ func (m *Worker) cancelDealTasks(deal *pb.Deal) error {
...
+	Networks  []*structs.NetworkSpec
+}
+
+type containerInfoMarshallable struct {
```
Comment (antmat) [inline insonmnia/worker/server.go]: why can't we introduce wrapper for Public key instead of containerInfo?

File: insonmnia/worker/server.go [hunk 4]
```diff
@@ -949,6 +970,11 @@ func (m *Worker) setupRunningContainers() error {
 				continue
 			}
 
+			info.Cinfo.ContainerInfo.PublicKey, _ = ssh.ParsePublicKey(info.Cinfo.PublicKey)
+
+			info.Description.Description.networks = info.Description.Networks
```
Comment (antmat) [inline insonmnia/worker/server.go]: looks weird

File: insonmnia/worker/server.go (hunk 1 above)
Comment (montekki) [reply insonmnia/worker/server.go]: Description struct has Networks() and Network() methods.

File: insonmnia/worker/server.go (hunk 4 above)
Comment (montekki) [reply insonmnia/worker/server.go]: it does, but yet in the existing code https://github.com/sonm-io/core/blob/master/insonmnia/worker/server.go#L601 ```ssh.ParsePublicKey()``` is not checked for returned error, as you may see. In the event when taskinfo has no ssh key that would be the case and containerInfo.PublicKey whill be ```nil```. So when we restore the task the same logic has to be applied: if no key is in the storage, containerInfo.PublicKey is also ```nil```. This logic definitely has to be fixed, but I don't that this PR is the place to do it.

File: insonmnia/worker/server.go (hunk 3 above)
Comment (montekki) [reply insonmnia/worker/server.go]: It is a matter of taste: some ugly wrapping and unwrapping will be done in any case. But when adding more fields to container info that have to have this ugly marshalling, you wouldn't have to add more new wrapper types for those fields.

File: insonmnia/worker/server.go (hunk 2 above)
Comment (montekki) [reply insonmnia/worker/server.go]: That would be handy if it helped getting rid of this whole new ugly ```descriptionMarshallable``` type, but it looks that the ```Networks``` field remains another obstacle in the way here, so in either way the type stays until the ```Networks\Networks()``` disambiguation is resolved.

**Pull request merged**